rekall-core/rekall/plugins/linux/heap_analysis.py

```diff
@@ -0,0 +1,4388 @@
+
+    for vma in vmas:
+        try:
+            print "{0}: {1:x}-{2:x}".format(vma['name'], vma['vma'].vm_start, vma['vma'].vm_end)
+        except UnicodeEncodeError:
+            print "{0}: {1:x}-{2:x}".format(repr(vma['name']),
+                                            vma['vma'].vm_start,
+                                            vma['vma'].vm_end)
+
+
+def get_libc_filename(vmas):
+    """Returns the libc file name from the vma, where the _LIBC_REGEX matches."""
+
+    if vmas:
+        for vma in vmas:
+            if re.search(_LIBC_REGEX, vma['name'], re.IGNORECASE):
+                return vma['name']
+
+
+def get_libc_range(vmas):
+    """Returns the lowest and highest address for the libc vma. See also get_mem_range_for_regex.
+    """
+
+    return get_mem_range_for_regex(vmas, _LIBC_REGEX)
+
+
+def get_mem_range_for_regex(vmas, regex):
+    """Returns the lowest and highest address of memory areas belonging to the vm_areas,
+    the given regex matches on.
+    The result is given as a list, where the lowest address is the first element.
+    Expects the output from _get_vmas_for_task as argument.
+    """
+
+    offsets = None
+    if vmas:
+        for vma in vmas:
+            if re.search(regex, vma['name'], re.IGNORECASE):
+                if not offsets:
+                    offsets = [vma['vma'].vm_start]
+                    offsets.append(vma['vma'].vm_end)
+
+                else:
+                    offsets[1] = vma['vma'].vm_end
+
+    return offsets
+
+
+
+class HeapAnalysis(common.LinProcessFilter):
+    """Basic abstract class for linux heap analysis.
+        Mostly serves the main_arena.
+    """
+
+    __abstract = True
+
+    _main_heap_identifier = "[heap]"
+
+    # used to mark vm_areas residing between the [heap] and the first file or stack object
+    # and all other vm_areas that have no file object for (mmapped regions can also reside somewhere
+    # beyond the typical heap/stack area)
+    # note that vmas with this identifier might be empty, old thread stacks or vm_areas belonging
+    # to e.g. mapped files
+    # a pretty reliable source for heap vm_areas is the self.heap_vmas list as it contains
+    # vm_areas which are identified to most probably belong to a heap or mmapped region
+    _heap_vma_identifier = "[heap-vma]"
+    _pot_mmapped_vma_identifier = "[pot-mmapped-vma]"
+
+
+    # is normally only automatically set when using a dummy arena or the chunk dumper, as in those
+    # cases all chunks are walked at least two times
+    def activate_chunk_preservation(self):
+        """Sets _preserve_chunks to True. This forces all allocated chunk functions to store
+        chunks in lists, which highly increases the speed of a second walk over those chunks.
+        This feature can only be activated if the command line option prevent_chunk_preservation
+        is not set.
+        """
+
+        if not self._preserve_chunks and not self._prevent_chunk_preservation:
```

Comment:
performance flag feels wrong in regards of "more thorough"
it's just a time/resources trade off
but i get your point
the plugin normally tries to choose the best solution, and to avoid too many dials, we could think about moving this feature to the performance flag